### PR TITLE
Single version source

### PR DESCRIPTION
**Pull request by lor11212**
JIRA: https://zrh-jira.esri.com/browse/CE-10315

This PR introduces a single source for the version of Puma in the file `version.properties`. The Rhino project, the Grasshopper project and the package creation script will all parse this file to get the version number.

**Implementation**

PumaRhino: Reads version.properties, parses each line to create `#define` macros, and write them to the `generatedVersion.h` file. This file is then used in version.h.

PumaGrasshopper: Reads and parses version.properties to create a version string and assign it to `AssemblyVersion` property.

Package creation: I added a function that reads and parses `version.properties`, and updates `manifest.yml` version line.

### CODE REVIEW

**Review by chr11115 — CHANGES_REQUESTED** [review]
Nice work! 
I'd suggest using template files with replacement tokens instead of "creating" the files from scratch through string manipulation.

File: PumaRhino/PumaRhino.vcxproj [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
-<Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project DefaultTargets="Build" InitialTargets="ReadVersion" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="ReadVersion">
+    <ReadLinesFromFile File="..\version.properties">
+      <Output TaskParameter="Lines" ItemName="PumaVersions" />
+    </ReadLinesFromFile>
+    <PropertyGroup>
+      <Version>#pragma once;@(PumaVersions-&gt;Replace('=', ' ')-&gt;'#define %(FileName)')</Version>
+    </PropertyGroup>
+	<WriteLinesToFile File=".\generatedVersion.h" Lines="$(Version)" Overwrite="true"></WriteLinesToFile>
```
[the remark this answers is not in the document]
Comment (chr11115) [reply PumaRhino/PumaRhino.vcxproj:10]: Update: 
It appears I linked a community version of msbuild. If msbuild does not offer similar functionality you can try to find an alternative way to fill a template file or leave it if it does not seem sensible.

**Comment by lor11212** [comment]
I incorporated your feedback 👍

File: PumaRhino/PumaRhino.vcxproj [hunk 2]
```diff
@@ -1,5 +1,33 @@
...
-<Project DefaultTargets="Build" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project DefaultTargets="Build" InitialTargets="ReadVersion" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="ReadVersion">
+    <ReadLinesFromFile File="..\version.properties">
+      <Output TaskParameter="Lines" ItemName="Versions" />
+    </ReadLinesFromFile>
+    <CreateItem Include="Versions" AdditionalMetadata="@(Versions)">
+      <Output TaskParameter="Include" ItemName="PumaVersions" />
+    </CreateItem>
+	<PropertyGroup>
```
Comment (chr11115) [inline PumaRhino/PumaRhino.vcxproj:10]: Mixed tabs/spaces

File: PumaGrasshopper/PumaGrasshopper.csproj [hunk 3]
```diff
@@ -1,5 +1,33 @@
...
+      <VersionMinor>%(PumaVersions.VERSION_MINOR)</VersionMinor>
+      <VersionBuild>%(PumaVersions.VERSION_BUILD)</VersionBuild>
+      <VersionRevision>%(PumaVersions.VERSION_REVISION)</VersionRevision>
+    </PropertyGroup>
+    <ReadLinesFromFile File=".\Properties\AssemblyVersion.template">
+      <Output TaskParameter="Lines" ItemName="TemplateVersionStrings" />
+    </ReadLinesFromFile>
+    <PropertyGroup>
+      <GeneratedVersionStrings>
+			  @(TemplateVersionStrings-&gt;
```
Comment (chr11115) [inline PumaGrasshopper/PumaGrasshopper.csproj:21]: mixed tabs/spaces

**Review by chr11115 — CHANGES_REQUESTED** [review]
Good work! 
Found some minor formatting nags. 
Did a quick functional test and I noticed that a clean build is necessary to adapt any version changes made. 

I don't know the best approach to fix this, but adding the lines below to the targets in PumaGrasshopper.csproj/PumaRhino.vcxproj fixed the issue for me.
```
<PropertyGroup>
    <DisableFastUpToDateCheck>true</DisableFastUpToDateCheck>
</PropertyGroup>
```
(https://stackoverflow.com/a/41803340)

**Review by chr11115 — APPROVED** [review]
LGTM, I decided to remove the hack to force build the version file. I guess it's better to require a force build every version change rather than forcing ms build tools over visual studio heuristics.. 
> A boolean value that applies to Visual Studio only. The Visual Studio build manager uses a process called FastUpToDateCheck to determine whether a project must be rebuilt to be up to date. This process is faster than using MSBuild to determine this. Setting the DisableFastUpToDateCheck property to true lets you bypass the Visual Studio build manager and force it to use MSBuild to determine whether the project is up to date.
(see https://learn.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-properties?view=vs-2022)

**Pull request merged**